core/sql/executor/HBaseClient_JNI.cpp

```diff
@@ -2844,819 +2871,980 @@ HBC_RetCode HBaseClient_JNI::getHbaseTableInfo(const char* tblName,
-    JavaMethods_[JM_DIRECT_CHECKANDINSERT     ].jm_signature = "(J[BLjava/lang/Object;JZ)Z";
-    JavaMethods_[JM_DIRECT_INSERT_ROWS ].jm_name      = "putRows";
-    JavaMethods_[JM_DIRECT_INSERT_ROWS ].jm_signature = "(JSLjava/lang/Object;Ljava/lang/Object;JZZ)Z";
-    JavaMethods_[JM_DIRECT_DELETE_ROWS ].jm_name      = "deleteRows";
-    JavaMethods_[JM_DIRECT_DELETE_ROWS ].jm_signature = "(JSLjava/lang/Object;J)Z";
-    JavaMethods_[JM_FETCH_ROWS ].jm_name      = "fetchRows";
-    JavaMethods_[JM_FETCH_ROWS ].jm_signature = "()I";
-    JavaMethods_[JM_DIRECT_GET_ROWS ].jm_name      = "getRows";
-    JavaMethods_[JM_DIRECT_GET_ROWS ].jm_signature = "(JSLjava/lang/Object;[Ljava/lang/Object;)I";
-    JavaMethods_[JM_COMPLETE_PUT ].jm_name      = "completeAsyncOperation";
-    JavaMethods_[JM_COMPLETE_PUT ].jm_signature = "(I[Z)Z";
-   
-    rc = (HTC_RetCode)JavaObjectInterface::init(className, javaClass_, JavaMethods_, (Int32)JM_LAST, javaMethodsInitialized_);
-    javaMethodsInitialized_ = TRUE;
-    pthread_mutex_unlock(&javaMethodsInitMutex_);
+  jobject jRow = jenv_->NewDirectByteBuffer(row.val, row.len);
+  if (jRow == NULL) {
+    GetCliGlobals()->setJniErrorStr(getErrorText(HBC_ERROR_INSERTROW_PARAM));
+    if (htc != NULL)
+        NADELETE(htc, HTableClient_JNI, heap);
+    jenv_->PopLocalFrame(NULL);
+    return HBC_ERROR_INSERTROW_PARAM;
   }
-  return rc;
-}
-
-//////////////////////////////////////////////////////////////////////////////
-// 
-//////////////////////////////////////////////////////////////////////////////
-NAString HTableClient_JNI::getLastJavaError()
-{
-  return JavaObjectInterface::getLastJavaError(JavaMethods_[JM_GET_ERROR].methodID);
+  jlong j_htc = (long)htc;
+  jboolean j_useTRex = useTRex;
+  jlong j_tid = transID;  
+  jlong j_ts = timestamp;
+  jboolean j_checkAndPut = checkAndPut;
+  jboolean j_asyncOperation = asyncOperation;
+ 
+  if (hbs)
+    hbs->getTimer().start();
+  tsRecentJMFromJNI = JavaMethods_[JM_HBC_DIRECT_INSERT_ROW].jm_full_name;
+  jboolean jresult = jenv_->CallBooleanMethod(javaObj_, JavaMethods_[JM_HBC_DIRECT_INSERT_ROW].methodID, 
+               	j_htc, js_tblName, j_useTRex, j_tid, jba_rowID, jRow, j_ts, j_checkAndPut, j_asyncOperation);
+  if (hbs) {
+      hbs->incHbaseCalls();
+      if (!asyncOperation)
+         hbs->incMaxHbaseIOTime(hbs->getTimer().stop());
+  }
+  if (jenv_->ExceptionCheck()) {
+    getExceptionDetails();
+    logError(CAT_SQL_HBASE, __FILE__, __LINE__);
+    logError(CAT_SQL_HBASE, "HBaseClient_JNI::insertRow()", getLastError());
+    jenv_->PopLocalFrame(NULL);
+    if (htc != NULL)
+        NADELETE(htc, HTableClient_JNI, heap);
+    return HBC_ERROR_INSERTROW_EXCEPTION;
+  }
+  if (hbs)
+    hbs->incBytesRead(rowID.len + row.len);
+  if (jresult == false) {
+     jenv_->PopLocalFrame(NULL);
+    if (htc != NULL)
+        NADELETE(htc, HTableClient_JNI, heap);
+     return HBC_ERROR_INSERTROW_DUP_ROWID;
+  }
+  *outHtc = htc;
```

Comment:
Yes. htc is null for sync operations. This asyncHtc_ in ExpHbaseInterface_JNI will be set to null. Hence, there is no need for release the asyncHtc_